### PR TITLE
Sg/swizzle node UI alignment 2

### PR DESCRIPTION
**Pull request by tracychen1234**
# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [x] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [x] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?
Update current Swizzle node UI for better alignment with other graph tools.
image

---
### Testing status
Describe what manual/automated tests were performed for this PR
Tested with different vector size input values and mask input edge cases.
Succeed in both local Paly Mode/Edit Mode tests.

---
### Comments to reviewers
Notes for the reviewers you have assigned.
I wrongly reverted a merge commit in the previous sg/swizzle-node-ui-alignment [PR](https://github.com/Unity-Technologies/Graphics/pull/2897), which caused many change files and broke the branch. This is the new branch & PR with the intended changes. Apologies for any inconvenience caused!

### CODE REVIEW

**Review by LandonTownsendUnity — APPROVED** [review]
All my concerns have been addressed. Did a simple upgrade test since just to make sure that still worked and approving.

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 1]
```diff
@@ -27,146 +26,177 @@ public override bool hasPreview
...
+                char[] rgba = { 'r', 'g', 'b', 'a' };
 
-                m_AlphaChannel = value;
-                Dirty(ModificationScope.Node);
+                for (int i = 0; i < MaskChars.Length; i++)
+                {
+                    if (rgba.Contains(MaskChars[i]))
+                    {
+                        MaskChars[i] = mask_map[MaskChars[i]];
+
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:90]: empty line, could be removed

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 2]
```diff
@@ -27,146 +26,177 @@ public override bool hasPreview
...
-                    s_ComponentList[m_AlphaChannel].ToString(CultureInfo.InvariantCulture));
+            var InputValueSize = SlotValueHelper.GetChannelCount(inputValueType);
+
+            if (!ValidateMaskInput(InputValueSize))
+            {
+                sb.AppendLine(string.Format("{0} {1} = 0;", outputSlotType, outputName));
+            }
             else
-                sb.AppendLine("{0} {1} = {0}({3}[((int){2} >> 0) & 3], {3}[((int){2} >> 2) & 3], {3}[((int){2} >> 4) & 3], {3}[((int){2} >> 6) & 3]);", outputSlotType, outputName, GetVariableNameForNode(), inputValue);
+            {
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:132]: This is great! Much simpler than whatever was happening before, bitshifting is scary 😅

File: com.unity.shadergraph/Editor/Drawing/Controls/TextControl.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                }
+                this.MarkDirtyRepaint();
+            });
+            field.Q("unity-text-input").RegisterCallback<FocusOutEvent>(evt =>
+            {
+                //Only set node value when mouse clicked away
+                m_Node.owner.owner.RegisterCompleteObjectUndo("Change" + m_Node.name);
+                m_PropertyInfo.SetValue(m_Node, value, null);
+                m_UndoGroup = -1;
+                //Validate graph to update downstream input slot
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/Drawing/Controls/TextControl.cs:72]: If possible I'd make it so that there was an Action/Delegate here that triggered the responses of whichever node owns this TextControl. This way, every node can implement its own response to the value being changed. 

Might be a bit convoluted because this is all implemented as an Attribute, maybe could have another attribute for a method that defines the value change response? 

`TextControlValueChangedAttribute` or something like that maybe, just an idea, there are probably better ways to do it.

It's not a huge problem right now as this is only used by the SwizzleNode, but thinking about how this might need to scale to other nodes over time and all that.

**Review by Nightmask3 — APPROVED** [review]
Left some comments, like the new TextControl, can definitely see the performance impact of dirty-ing the nodes every frame of the input being entered so this is much better, good solution! 👍 

Some things that could be addressed before merging in but nothing that's a huge blocker or issue. Approved!

File: com.unity.shadergraph/Editor/Drawing/Controls/TextControl.cs (hunk 3 above)
Comment (Nightmask3) [reply com.unity.shadergraph/Editor/Drawing/Controls/TextControl.cs:72]: Had a discussion with Tracy offline and I think we're justified in keeping this in for now and looking at it later. The Swizzle node is the only user of the TextControlView in the graph at the moment and all of this code is likely to drastically change with GTF down the line so its not worth spending more time on what was a usability week task.

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 4]
```diff
@@ -27,146 +26,174 @@ public override bool hasPreview
...
-            if ((int)greenChannel >= channelCount)
-                greenChannel = TextureChannel.Red;
-            if ((int)blueChannel >= channelCount)
-                blueChannel = TextureChannel.Red;
-            if ((int)alphaChannel >= channelCount)
-                alphaChannel = TextureChannel.Red;
+            AddSlot(new DynamicVectorMaterialSlot(InputSlotId, kInputSlotName, kInputSlotName, SlotType.Input, Vector4.zero));
+            switch (_maskInput.Length)
+            {
+                default:
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:102]: This switch statement looks backwards to me, and it makes me stop and wonder if there's a reason you did it that way (usually numbers go in ascending order and default comes last). Is there a reason?

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 5]
```diff
@@ -27,146 +26,174 @@ public override bool hasPreview
...
+            var InputValueSize = SlotValueHelper.GetChannelCount(inputValueType);
+
+            if (!ValidateMaskInput(InputValueSize))
+            {
+                sb.AppendLine(string.Format("{0} {1} = 0;", outputSlotType, outputName));
+            }
             else
-                sb.AppendLine("{0} {1} = {0}({3}[((int){2} >> 0) & 3], {3}[((int){2} >> 2) & 3], {3}[((int){2} >> 4) & 3], {3}[((int){2} >> 6) & 3]);", outputSlotType, outputName, GetVariableNameForNode(), inputValue);
+            {
+                string outputValue = "";
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:132]: Why copy the characters one by one into a new string? Can't you just use convertedMask directly in the AppendLine call below?

File: com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs [hunk 6]
```diff
@@ -27,146 +26,174 @@ public override bool hasPreview
...
-        [ChannelEnumControl("Green Out")]
-        public TextureChannel greenChannel
+        public bool ValidateMaskInput(int InputValueSize)
         {
-            get { return m_GreenChannel; }
-            set
-            {
-                if (m_GreenChannel == value)
-                    return;
+            _maskInput = _maskInput.ToLower();
```
Comment (marctem) [inline com.unity.shadergraph/Editor/Data/Nodes/Channel/SwizzleNode.cs:50]: I like the ToLower() call since it lets users not care - artists hate case sensitivity. However, this overwrites their chosen case when they save and reload. Can we keep the lower case version inside our own code (just for validation and actually writing the shader)?

**Review by marctem — CHANGES_REQUESTED** [review]
A few suggestions, nothing major. Good job!

**Comment by ValGrimm-U3D** [comment]
Adjusted longer version of text Tracy adjusted. Good adjustments! Sent for editorial review.

**Review by ValGrimm-U3D — APPROVED** [review]
Will PR and push any editorial review changes to doc after merge.

I have some proposed doc edits if you'd like them from the Google Doc we were collaborating on, but otherwise the doc can just be adjusted later after editorial has given their input. 

 https://docs.google.com/document/d/1408EZyTD9lQR9dVlvvfbPpSAo3R_SA83bIktkZfXqNk/edit

Proposed revision:
Reorders the elements of the input vector as you specify in a formatting string in the input mask. For example, "wzyx" and "abgr" both invert the order of the input elements. The length of the input mask also determines the dimensions of the output vector. To output a vector3 with the x, y and z elements of the input vector, use the input mask “xyz”.
The error "Invalid Mask" indicates an input mask value which includes one or more channels that do not exist in the input vector.

**Comment by tracychen1234** [comment]
> Will PR and push any editorial review changes to doc after merge.
> 
> I have some proposed doc edits if you'd like them from the Google Doc we were collaborating on, but otherwise the doc can just be adjusted later after editorial has given their input.
> 
> https://docs.google.com/document/d/1408EZyTD9lQR9dVlvvfbPpSAo3R_SA83bIktkZfXqNk/edit
> 
> Proposed revision:
> Reorders the elements of the input vector as you specify in a formatting string in the input mask. For example, "wzyx" and "abgr" both invert the order of the input elements. The length of the input mask also determines the dimensions of the output vector. To output a vector3 with the x, y and z elements of the input vector, use the input mask “xyz”.
> The error "Invalid Mask" indicates an input mask value which includes one or more channels that do not exist in the input vector.

Updated swizzle-node.md :)

**Review by marctem — APPROVED** [review]
